cudax/include/cuda/experimental/__device/all_devices.cuh

```diff
@@ -0,0 +1,244 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDAX__DEVICE_ALL_DEVICES
+#define _CUDAX__DEVICE_ALL_DEVICES
+
+#include <cuda/__cccl_config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/std/__cccl/attributes.h>
+#include <cuda/std/__cuda/api_wrapper.h>
+#include <cuda/std/atomic>
+#include <cuda/std/cassert> // IWYU pragma: keep (for assert)
+#include <cuda/std/iterator> // IWYU pragma: keep (for random_access_iterator_tag)
+
+#include <cuda/experimental/__device/device.cuh>
+
+namespace cuda::experimental
+{
+namespace __detail
+{
+//! @brief A random-access range of all available CUDA devices
+class all_devices
+{
+public:
+  using size_type = _CUDA_VSTD::size_t;
+  struct iterator;
+  using const_iterator = iterator;
+
+  all_devices() = default;
+
+  _CCCL_NODISCARD constexpr device operator[](size_type __i) const noexcept;
+
+  _CCCL_NODISCARD size_type size() const;
+
+  _CCCL_NODISCARD iterator begin() const noexcept;
+
+  _CCCL_NODISCARD iterator end() const noexcept;
```

Comment:
Suggestion: I first wondered why we just don't have a global `std::vector<..>`, for the sake of having a familiar interface. However, we would not need to resize, or even allocate at all, so a custom type seems fair. Still, I miss some familiar member functions, like `at(i)` and `front()`. I could image people wanting to write quick & dirty code like:
```c++
int main () {
  auto dev = cuda::devices.at(0); // or front()
  ...
}
```
With their program being fine to just throw and die when there is no CUDA device.

Reply:
Since @miscco approved the PR, I would like to ping @ericniebler on this suggestion again. I would like to have at least `at(i)`, which throws when out of bounds.

Reply:
Done.

Reply:
Thx!